apply ceil() to canvas (prevent potential resampling from fractional pixel dimensions)

bcairns opened the pull request:
https://github.com/google/model-viewer/issues/3554

Fractional pixel dimensions (eg, 650.8) on canvas can cause "1px upsampling" (canvas is scaled larger by a single pixel), causing a slight "resampling blur" compared to native render size.

This can be fixed by removing the fractional parts of the dimensions.

After an initial solution using floor(), we are now switching to using ceil() to eliminate 1px gap issues.  (Concerns about three.js using floor() are moot; we just feed it integers anyway)

### Notes

- As we are already focused on canvas dimensions here, we are also attempting to resolve 1px gap issues (in addition to initial blurring issue).  This is accomplished by using ceil() and hiding overflow
- Only "final" values are adjusted (retain floating point until the last step)
- Tested on both DPR 1 & 2 screens, and at different browser zoom levels, with both my real-world use case and a test page with a variety of models displayed at different sizes (on Chrome, Firefox, and Safari)
- We don't need to apply ceil() to .container; in fact this is bad because it just creates another potential gap on ITS container
- I seemed to need overflow:hidden on .container, I noticed some 1px overlap weirdness without it (the canvas would overlap a 1px border on &lt;model-viewer>)... perhaps it works better because .container has width/height "set in stone" with inline styling values? (Always hard to say for certain when "how browsers round things" is part of the equation)  I then removed overflow:hidden from .userInput as redundant
- Ideas for further improvement: custom properties (aka CSS vars) might be useful for these dimensions, could be a bit more DRY

bcairns commented:
I did the CLA thing, I'm not sure why it says that is an issue

elalish reviewed (commented):
Thank you! I'm curious about the 1px visible line thing too. So far any that I've found on our dev pages also show up on the master version. If you notice any that are different, please share!

bcairns commented on packages/model-viewer/src/three-components/Renderer.ts:
```diff
@@ -246,8 +246,8 @@ export class Renderer extends EventDispatcher {
-    const width = this.width / scale;
-    const height = this.height / scale;
+    const width = Math.floor(this.width / scale);
+    const height = Math.floor(this.height / scale);
```
Ooh good spot, I missed that.

_Potentially_ it (lines ~475) is moot, because it's using Math.min() with this.canvas3D dimensions? (which I think get floored).

But not 100% sure and I can't really check thoroughly right now; I am on vacation and won't be able to investigate/test this properly until end of next week.

bcairns replied:
Hi there, I have also applied floor() in the render function.

In my specific testing (my real-world use case with fractional pixels values) I observed no difference here between using ceil() and floor() at all; they were always exactly the same as the (integer) this.canvas3D.width / height values.  (it's possible this Math.min() logic is no longer actually needed at all? but I don't know the code well enough to have any confidence there)

Anyway, for immediate purposes it seems prudent to be consistent and use floor here.

bcairns replied:
Oh interesting, thanks for explaining.

I have created a local test page displaying 3 different models at different sizes (and with a mix of integer/floats for container sizes).

It appears to be working as per your above description; the <canvas> elements take the max width/height across all instances, and actual renders appear at correct dimensions.

elalish reviewed (approved):
Thank you for noticing the fine details and taking the time to fix them! I really appreciate your testing.

bcairns commented:
> Thank you for noticing the fine details and taking the time to fix them! I really appreciate your testing.

You are very welcome, happy I was able to contribute.

So what happens now? It says full_test_run is still needed. Do I need to do anything else on my end?

And if I may ask, any idea on the general timing of when this fix might make its way into a release?

bcairns commented:
> Thank you! I'm curious about the 1px visible line thing too. So far any that I've found on our dev pages also show up on the master version. If you notice any that are different, please share!

Fractional canvas dimensions that would be rounded upwards (> x.5) will exhibit a 1px gap between the model-viewer and its container.

Here is a specific example where a gap would be present inside the 1px black border:

```
<div class="container">
  <model-viewer ... />
</div>

<style>
.container {
  width: 456.78px;
  height: 234.56px
  border: 1px solid black;
}
model-viewer {
  width: height: 100%;
  height: 100%;
}
</style>
```

The actual render will be @ 456.0 x 234.0, whereas the browser will display .container @ 457.0 x 235.0 effective pixels (subject to DPR considerations... ie with DPR 2 you'd wind up with 457.0 x 234.5 effective pixels for .container)

elalish commented:
Interesting; does this mean it would be better to use `ceil` than `floor` so it can be bigger and the overflow can be hidden? Or does that introduce other problems? 

As for timing, not sure. Hopefully inside of a month?

bcairns commented:
> Interesting; does this mean it would be better to use `ceil` than `floor` so it can be bigger and the overflow can be hidden? Or does that introduce other problems?

Yes I think ceil() combined with overflow:hidden on &ltmodel-viewer> (ie :host) itself should work.  It's something I could attempt.

My knee-jerk reaction is "ahh! truncating pixels is bad!", but it's not something that is actually noticeable, while a gap certainly can be.

My original rationale behind using floor() was because it appears that three.js uses floor() exclusively, and aligning with that seemed like the safest choice. (eg: https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L415)

"Ideally" (for certain notions of ideal), I would think round() that takes DPR into consideration would be best in theory.  So with DPR 2 you could "round" to 0.5 pixels, etc.  So like Math.round( value * dpr ) / dpr.  I _suspect_ that would match up best with how browsers round things off to "closest effective pixel".  But that feels like opening a new can of worms and I don't really think I can tackle that right now.  And ceil() should be perfectly good and much simpler.

**tl;dr** I'll take a crack at using ceil() + overflow:hidden and see if I can eliminate the gap.

elalish commented on packages/model-viewer/src/model-viewer-base.ts:
```diff
@@ -518,10 +518,13 @@ export default class ModelViewerElementBase extends ReactiveElement {
    */
   [$updateSize]({width, height}: {width: any, height: any}) {
+    width = Math.floor(parseFloat(width));
+    height = Math.floor(parseFloat(height));
```
Your last comment got me thinking. I believe you're right and we should change these to `width = Math.ceil(parseFloat(width) * dpr) / dpr;`

bcairns replied:
This code is now actually left untouched (.container retaining floating point dimensions appears to be a better solution in conjunction with the new ceil() approach).

bcairns commented:
OK, commit #3 in the increasing inaccurately named "canvas-floor" branch.

I think I've got it working really well now.

Bit of a lightbulb moment; now that we're going ceil() and clipping, we don't need to round off .container's dimensions at all.  In fact doing so is bad because it actually just "kicks the can up the road" and creates another potential gap on one of ITS containers.

So .container stays floating point and occupies the full amount it should, canvas renders at the right (ceil'd) size, and everyone's happy.  Blur issue is resolved and I can't produce any more 1px gaps.

I updated the original PR description above.  Please also see notes about why I moved overflow:hidden from .userInput to .container. (Let me know if there are any concerns there)

elalish reviewed (approved):
This looks great; thank you for the thorough testing!

bcairns commented:
> This looks great; thank you for the thorough testing!

My pleasure, thanks for your insights in evolving the solution. I'm happy it looks like we can kill two birds with one stone here (knock on wood).
